### ISSUE TITLE
Mismatch between execution_space tag in Functor and execution space in execution policy is silently ignored

### ISSUE THREAD

**Issue by dhollman** [kokkos/kokkos#2074]
This program should have a compilation error:

```c++
using cuda_exec_policy = Kokkos::TeamPolicy<Kokkos::Cuda>;
using serial_exec_policy = Kokkos::TeamPolicy<Kokkos::Serial>;

template <class ExecSpace>
struct MyFunctor {
  using execution_space = ExecSpace;
  KOKKOS_INLINE_FUNCTION
  void operator()(typename serial_exec_policy::member_type const& i) const {
    printf("running Serial version\n");
  }
  KOKKOS_INLINE_FUNCTION
  void operator()(typename cuda_exec_policy::member_type const& i) const {
    printf("running Cuda version\n");
  }
};

int main(int argc , char* argv[]) {

  Kokkos::initialize(argc,argv);

  {
    Kokkos::parallel_for(cuda_exec_policy{1, 1}, MyFunctor<Kokkos::Serial>{});
    Kokkos::fence();

    Kokkos::parallel_for(serial_exec_policy{1, 1}, MyFunctor<Kokkos::Cuda>{});
    Kokkos::fence();
  }

  Kokkos::finalize();

}
```

But it actually just compiles, runs, and prints: 

```
running Cuda version
running Serial version
```

In other words, the functor's `execution_space` attestation is just silently ignored.

**Comment by dhollman** [comment]
I plan to fix this a part of #2034

**Comment by ldh4** [comment]
This problem is still there. The `execution_space` of a functor is only referenced when internally constructing an `execution_policy (RangePolicy by default)`, but if an `execution_policy` has been provided, the `execution_space` of the functor seems to be unused.

**Comment by ajpowelsnl** [comment]
Hi @ldh4 -- How's this issue going?  Do you have an ETA for completion?

**Comment by ldh4** [comment]
@ajpowelsnl I believe #4655 resolved this issue.

**Comment by ajpowelsnl** [comment]
> @ajpowelsnl I believe #4655 resolved this issue.

OK, cool.  Would you please verify this, and if #4655 does indeed provide the fix, would you please close this issue?

**Comment by masterleinad** [comment]
> OK, cool. Would you please verify this, and if #4655 does indeed provide the fix, would you please close this issue?

@ajpowelsnl This issue has the `InDevelop` tag. It's not necessary to reverify following our way to deal with issues (If the issue wasn't introduced in `develop` mark it as `InDevelop` as soon as resolved in `develop` and close it once it is in `master`. If the issue was introduced in`develop` close it as soon as a fix is merged to `develop`).

**Comment by PhilMiller** [comment]
Fix released in 3.6.01
